File: 22. Generate Parentheses/22.go
package leetcode

func generateParenthesis(n int) []string {
    if n == 0{
		return []string{}
	}

	res := make([]string, 0)

	r := n
	l := 0

	dfs(&res, "", 2*n, r, l)

	return res
}

func dfs(res *[]string, tmpStr string, n, r, l int){
	if len(tmpStr) == n{
		*res = append(*res, tmpStr)
	}

	if r != 0{
		dfs(res, tmpStr+"(", n, r-1, l+1)
	}
	if l != 0{
		dfs(res, tmpStr+")", n, r, l-1)
	}
}

File: 22. Generate Parentheses/_22.py
class Solution:
    def generateParenthesis(self, n: int):
        res = list()
        
        if n == 0:
            return res
        
        l = 0
        r = n
        self.dfs(res, "", r, l, 2*n)

        return res


    def dfs(self, res, tmpstr, r, l, n):
        if len(tmpstr) == n:
            res.append(tmpstr)

        if r != 0:
            self.dfs(res, tmpstr+'(', r-1, l+1, n)
        if l != 0:
            self.dfs(res, tmpstr+')', r, l-1, n)


a = Solution()
print(a.generateParenthesis(3))